src/SubscriberBase.h

```diff
@@ -18,34 +30,115 @@ class SubscriberBaseT : public Subscriber<T>,
+    // nullify subscription pointer after calling cancel. That is too strong of
+    // a requirement for the users.
+    std::weak_ptr<SubscriberBaseT<T>> parentSubscriber_;
+  };
+
+  friend class SubscriptionShim;
+
  public:
-  using ExecutorBase::ExecutorBase;
+  // in c++11 we have to declare this explicitly, instead of
+  // using ExecutorBase::ExecutorBase because of atomic cancelled :(
+  // maybe its gcc issue
+  explicit SubscriberBaseT(
+      folly::Executor& executor = defaultExecutor(),
+      bool startExecutor = true) : ExecutorBase(executor, startExecutor), cancelled_(false) {}
 
   void onSubscribe(std::shared_ptr<Subscription> subscription) override final {
     auto thisPtr = this->shared_from_this();
-    runInExecutor(
-        [thisPtr, subscription]() { thisPtr->onSubscribeImpl(subscription); });
+    runInExecutor([thisPtr, subscription]() {
+      CHECK(!thisPtr->cancelled_);
+      CHECK(!thisPtr->originalSubscription_);
+      thisPtr->originalSubscription_ = std::move(subscription);
+      thisPtr->onSubscribeImpl(
+          std::make_shared<SubscriptionShim>(thisPtr->shared_from_this()));
+    });
   }
 
   void onNext(T payload) override final {
     auto movedPayload = folly::makeMoveWrapper(std::move(payload));
     auto thisPtr = this->shared_from_this();
     runInExecutor([thisPtr, movedPayload]() mutable {
-      thisPtr->onNextImpl(movedPayload.move());
+      if (!thisPtr->cancelled_) {
+        thisPtr->onNextImpl(movedPayload.move());
+      }
     });
   }
 
   void onComplete() override final {
     auto thisPtr = this->shared_from_this();
-    runInExecutor([thisPtr]() { thisPtr->onCompleteImpl(); });
+    runInExecutor([thisPtr]() {
+      if (!thisPtr->cancelled_.exchange(true)) {
+        thisPtr->onCompleteImpl();
+
+        DCHECK(thisPtr->originalSubscription_);
+        thisPtr->originalSubscription_->cancel();
+        thisPtr->originalSubscription_ = nullptr;
+      }
+    });
   }
 
   void onError(folly::exception_wrapper ex) override final {
     auto movedEx = folly::makeMoveWrapper(std::move(ex));
     auto thisPtr = this->shared_from_this();
-    runInExecutor(
-        [thisPtr, movedEx]() mutable { thisPtr->onErrorImpl(movedEx.move()); });
+    runInExecutor([thisPtr, movedEx]() mutable {
+      if (!thisPtr->cancelled_.exchange(true)) {
+        thisPtr->onErrorImpl(movedEx.move());
+
+        DCHECK(thisPtr->originalSubscription_);
+        thisPtr->originalSubscription_->cancel();
+        thisPtr->originalSubscription_ = nullptr;
+      }
+    });
   }
+
+ private:
+  // Once the subscription is cancelled we will no longer deliver any
+  // other signals.
+  // Scheduling the calls is not atomic operation so it may very well happen
+  // that 2 threads race for sending onNext and onComplete. We need to make sure
+  // that once the terminating signal is delivered we no longer try to deliver
+  // onNext.
+  std::atomic<bool> cancelled_;
```

Comment:
This seems to only be used from the `Executor`, meaning this doesn't really need to be atomic.

(If it _is_ modified from elsewhere, there's a race in `onNext`.)